tests/test_global_settings.txt
bridge_skin_material_flow_3=110
speed_wall_0=20
infill_wipe_dist=0
machine_extruder_count=2
raft_surface_jerk=25
material_bed_temperature_layer_0=60
jerk_support_roof=5
roofing_pattern=lines
gradual_support_infill_step_height=1
wireframe_up_half_speed=0.3
support_angle=60
top_bottom_pattern_0=lines
support_bottom_extruder_nr=0
wall_line_width_x=0.3
skirt_brim_speed=20
raft_fan_speed=0
support_roof_enable=False
ooze_shield_dist=2
prime_blob_enable=True
acceleration_support=2000
machine_width=233
skirt_brim_line_width=0.35
raft_airgap=0.3
infill_overlap=0
magic_fuzzy_skin_point_dist=0.8
material_extrusion_cool_down_speed=0.7
cool_fan_speed=50
support_tree_collision_resolution=0.175
print_sequence=all_at_once
conical_overhang_angle=50
machine_show_variants=False
raft_base_jerk=25
wireframe_top_delay=0
wireframe_flat_delay=0.1
machine_acceleration=3000
remove_empty_first_layers=True
wireframe_fall_down=0.5
bridge_skin_density_3=80
platform_adhesion=0
acceleration_wall_x=1000
retraction_extra_prime_amount=0
speed_topbottom=20
support_roof_pattern=concentric
travel=0
alternate_carve_order=True
adhesion_type=brim
bridge_skin_density=100
raft_base_thickness=0.2412
bridge_settings_enabled=False
support_bottom_stair_step_width=5.0
support_tree_branch_diameter_angle=5
jerk_topbottom=5
bridge_skin_speed_2=10.0
wireframe_roof_fall_down=2
raft_smoothing=5
support=0
cutting_mesh=False
relative_extrusion=False
support_fan_enable=False
support_infill_rate=0
bridge_skin_material_flow=60
infill_wall_line_count=0
experimental=0
machine_center_is_zero=False
infill=0
infill_enable_travel_optimization=False
raft_base_acceleration=4000
switch_extruder_prime_speed=15
speed_travel=250
support_minimal_diameter=3.0
acceleration_support_interface=500
material_flow=100
connect_skin_polygons=False
raft_base_line_spacing=1.6
support_bottom_distance=0.2
wall_thickness=1
magic_fuzzy_skin_enabled=False
raft_surface_thickness=0.2
material_print_temperature=200
raft_interface_line_spacing=0.8999999999999999
wall_0_wipe_dist=0.2
support_roof_height=1
support_conical_angle=30
speed_roofing=20
anti_overhang_mesh=False
meshfix_union_all_remove_holes=False
wireframe_roof_inset=3
wireframe_drag_along=0.6
wireframe_flow_flat=100
support_interface_skip_height=0.3
draft_shield_height=10
top_thickness=1
machine_endstop_positive_direction_y=False
adaptive_layer_height_enabled=False
skin_line_width=0.35
speed_support_bottom=23
raft_surface_acceleration=4000
carve_multiple_volumes=True
start_layers_at_same_position=False
coasting_enable=False
filter_out_tiny_gaps=True
support_bottom_line_width=0.35
support_conical_enabled=False
mesh_position_x=0
raft_surface_speed=17.5
material_adhesion_tendency=10
bridge_fan_speed_3=0
draft_shield_enabled=False
switch_extruder_retraction_speed=20
smooth_spiralized_contours=True
raft_interface_line_width=0.7
connect_infill_polygons=False
top_skin_expand_distance=0.95
speed_prime_tower=20
retraction_min_travel=5
infill_pattern=lines
material_print_temp_prepend=True
wall_line_width=0.35
machine_name=Ultimaker 3
retraction_hop_enabled=True
fill_perimeter_gaps=everywhere
z_seam_y=645
meshfix_extensive_stitching=False
ironing_enabled=False
jerk_wall=10
bridge_skin_speed=10.0
top_bottom_thickness=1
acceleration_support_bottom=500
support_roof_line_distance=0.35
wall_line_width_0=0.35
machine_min_cool_heat_time_window=15
support_interface_pattern=concentric
zig_zaggify_infill=True
support_tree_enable=True
roofing_layer_count=0
meshfix_keep_open_polygons=False
support_interface_extruder_nr=0
wall_0_extruder_nr=-1
bridge_wall_min_length=5
support_mesh_drop_down=True
wall_line_count=3
retract_at_layer_change=False
machine_steps_per_mm_y=50
machine_filament_park_distance=42.5
skin_outline_count=1
retraction_count_max=10
material_guid=
acceleration_infill=4000
time=14:47:24
layer_0_z_overlap=0.15
speed_support_infill=20
support_interface_line_width=0.35
mold_roof_height=0.5
machine_heat_zone_length=16
support_tree_angle=40
support_bottom_stair_step_height=0.3
meshfix_maximum_resolution=0.04
raft_interface_fan_speed=0
machine_nozzle_id=unknown
retraction_speed=25
jerk_layer_0=5
support_interface_enable=False
speed_equalize_flow_max=150
cool_min_speed=5
magic_mesh_surface_mode=normal
acceleration_roofing=500
infill_sparse_thickness=0.2
wireframe_strategy=compensate
cross_infill_density_image=
z_seam_relative=False
raft_surface_fan_speed=0
fill_outline_gaps=False
support_join_distance=2.0
machine_nozzle_heat_up_speed=1.4
support_line_width=0.35
bridge_enable_more_layers=True
infill_mesh_order=0
retraction_hop=2
speed_layer_0=20
prime_tower_line_width=0.35
ironing_line_spacing=0.1
jerk_print_layer_0=5
dual=0
machine_max_feedrate_x=300
max_skin_angle_for_expansion=90
jerk_infill=25
machine_endstop_positive_direction_z=True
bridge_fan_speed_2=0
wall_extruder_nr=-1
retraction_amount=6.5
travel_avoid_distance=3
coasting_speed=90
support_z_distance=0.4
blackmagic=0
raft_surface_layers=2
gantry_height=60
raft_interface_jerk=25
acceleration_topbottom=500
infill_sparse_density=100
wireframe_printspeed_flat=5
speed_wall=30
skin_overlap=10
jerk_skirt_brim=5
limit_support_retractions=True
switch_extruder_retraction_speeds=20
skirt_brim_minimal_length=250
cool_fan_speed_min=50
machine_disallowed_areas=[[[92.8, -53.4], [92.8, -97.5], [116.5, -97.5], [116.5, -53.4]], [[73.8, 107.5], [73.8, 100.5], [116.5, 100.5], [116.5, 107.5]], [[74.6, 107.5], [74.6, 100.5], [116.5, 100.5], [116.5, 107.5]], [[74.9, -97.5], [74.9, -107.5], [116.5, -107.5], [116.5, -97.5]], [[-116.5, -103.5], [-116.5, -107.5], [-100.9, -107.5], [-100.9, -103.5]], [[-116.5, 105.8], [-96.9, 105.8], [-96.9, 107.5], [-116.5, 107.5]]]
draft_shield_height_limitation=full
travel_retract_before_outer_wall=True
date=30-08-2018
travel_compensate_overlapping_walls_enabled=True
support_offset=0.2
wireframe_enabled=False
support_use_towers=True
mold_angle=40
infill_line_width=0.42
prime_tower_min_volume=6
bridge_skin_speed_3=10.0
bottom_skin_preshrink=0.95
gradual_support_infill_steps=0
top_bottom_pattern=lines
acceleration_ironing=500
spaghetti_infill_enabled=False
conical_overhang_enabled=False
material_standby_temperature=100
skin_overlap_mm=0.032499999999999994
raft_base_fan_speed=0
raft_base_line_width=0.8
mesh_rotation_matrix=[[1,0,0], [0,1,0], [0,0,1]]
support_pattern=zigzag
cooling=0
wireframe_printspeed_bottom=5
support_roof_line_width=0.35
skirt_gap=3
wireframe_top_jump=0.6
speed_support=20
acceleration_print=4000
speed_wall_x=30
support_xy_distance_overhang=0.35
speed_equalize_flow_enabled=False
material_bed_temp_wait=True
acceleration_print_layer_0=500
material_surface_energy=100
speed_print_layer_0=20
brim_outside_only=True
support_tower_roof_angle=65
extruder_prime_pos_x=0
meshfix_union_all=True
machine_head_with_fans_polygon=[[-41.9, -45.8], [-41.9, 33.9], [59.9, 33.9], [59.9, -45.8]]
command_line_settings=0
z_seam_corner=z_seam_corner_inner
machine_steps_per_mm_x=50
default_material_bed_temperature=60
cool_min_layer_time=5
retraction_prime_speed=15
flow_rate_max_extrusion_offset=0
magic_fuzzy_skin_point_density=1.25
material_final_print_temperature=185
spaghetti_max_height=2.0
coasting_volume=0.064
skirt_line_count=1
machine_max_feedrate_e=45
wireframe_height=3
acceleration_prime_tower=2000
machine_gcode_flavor=Griffin
machine_height=200
jerk_travel=30
machine_max_feedrate_y=300
prime_tower_position_y=184.56
machine_max_feedrate_z=40
infill_extruder_nr=1
infill_before_walls=True
resolution=0
mold_width=5
line_width=0.35
raft_surface_line_width=0.35
jerk_ironing=5
retraction_combing_max_distance=0
acceleration_layer_0=500
coasting_min_volume=0.8
raft_margin=15
support_tower_diameter=3.0
wireframe_roof_outer_delay=0.2
cool_fan_speed_max=100
machine_endstop_positive_direction_x=False
speed_support_roof=23
ironing_only_highest_layer=False
support_conical_min_width=5.0
bridge_wall_coast=100
alternate_extra_perimeter=False
support_zag_skip_count=0
initial_extruder_nr=0
roofing_angles=[ ]
zig_zaggify_support=False
wireframe_nozzle_clearance=1
wireframe_printspeed=5
jerk_print=25
infill_offset_x=0
machine_max_acceleration_e=10000
bottom_skin_expand_distance=0.95
raft_interface_speed=13.125
acceleration_wall_0=500
spaghetti_max_infill_angle=10
support_bottom_line_distance=0.35
machine_nozzle_tip_outer_diameter=1
machine_nozzle_head_distance=3
prime_tower_flow=100
optimize_wall_printing_order=True
support_infill_extruder_nr=0
switch_extruder_retraction_amount=8
bottom_layers=999999
support_mesh=False
support_infill_sparse_thickness=0.2
expand_skins_expand_distance=0.95
retraction_combing=off
support_initial_layer_line_distance=0
support_enable=False
jerk_enabled=True
extruder_prime_pos_abs=True
z_seam_x=116.5
min_skin_width_for_expansion=0.0
print_bed_temperature=60
travel_avoid_other_parts=True
support_connect_zigzags=True
acceleration_support_roof=500
magic_fuzzy_skin_thickness=0.3
support_line_distance=0
acceleration_support_infill=2000
min_infill_area=0
bridge_skin_support_threshold=50
draft_shield_dist=10
machine_max_acceleration_x=9000
infill_offset_y=0
prime_tower_size=20
material_print_temp_wait=True
minimum_polygon_circumference=1.0
support_interface_density=100
infill_support_angle=40
support_roof_extruder_nr=0
machine_shape=rectangular
initial_layer_line_width_factor=120
default_material_print_temperature=200
infill_mesh=False
machine_nozzle_expansion_angle=45
skin_angles=[ ]
brim_line_count=17
machine_settings=0
support_extruder_nr_layer_0=0
meshfix_maximum_travel_resolution=0.2857142857142857
infill_angles=[ ]
ironing_inset=0.175
cool_fan_speed_0=0
machine_max_jerk_xy=20.0
max_feedrate_z_override=0
nozzle_disallowed_areas=[]
xy_offset_layer_0=0
machine_buildplate_type=glass
z_seam_type=sharpest_corner
jerk_prime_tower=15
extruder_prime_pos_y=0
brim_width=7
jerk_support=15
adhesion_extruder_nr=0
jerk_travel_layer_0=6.0
support_tree_branch_diameter=2
outer_inset_first=False
gradual_infill_steps=0
extruder_prime_pos_z=0
support_bottom_density=100
travel_compensate_overlapping_walls_0_enabled=True
support_bottom_enable=False
machine_max_acceleration_y=9000
raft_surface_line_spacing=0.35
bridge_wall_material_flow=50
wall_min_flow=0
cool_fan_full_at_height=1.0010000000000001
center_object=False
adaptive_layer_height_threshold=200.0
speed_travel_layer_0=142.85714285714286
support_bottom_height=1
travel_avoid_supports=False
speed_support_interface=20
machine_start_gcode=
prime_tower_position_x=175.70000000000002
infill_overlap_mm=0
multiple_mesh_overlap=0
machine_nozzle_cool_down_speed=0.8
prime_tower_wipe_enabled=False
prime_tower_enable=False
speed_print=35
raft_base_speed=13.125
retraction_retract_speed=25
machine_end_gcode=G91 ;Relative movement\nG0 F15000 X8.0 Z0.5 E-4.5 ;Wiping+material retraction\nG0 F10000 Z1.5 E4.5 ;Compensation for the retraction\nG90 ;Disable relative movement
support_tree_wall_count=1
wireframe_roof_drag_along=0.8
material_shrinkage_percentage=0
machine_steps_per_mm_e=1600
xy_offset=0
cross_support_density_image=
print_temperature=200
spaghetti_inset=0.2
spaghetti_infill_stepped=True
machine_max_acceleration_z=100
magic_spiralize=False
top_bottom_extruder_nr=-1
support_extruder_nr=0
support_xy_overrides_z=z_overrides_xy
material_flow_layer_0=100
day=Thu
bottom_thickness=1
wall_min_flow_retract=False
machine_use_extruder_offset_to_offset_coords=True

Changelog.md
Changelog PicasoXCore
====================

- [Not documented]

src/settings/settings.cpp
//Copyright (c) 2018 Ultimaker B.V.
//Copyright (c) 2020 PICASO 3D
//PicasoXCore is released under the terms of the AGPLv3 or higher

#include <cctype>
#include <fstream>
#include <stdio.h>
#include <sstream> // ostringstream
#include <regex> // regex parsing for temp flow graph
#include <string> //Parsing strings (stod, stoul).

#include "EnumSettings.h"
#include "FlowTempGraph.h"
#include "Settings.h"
#include "types/AngleDegrees.h" //For angle settings.
#include "types/AngleRadians.h" //For angle settings.
#include "types/Duration.h" //For duration and time settings.
#include "types/LayerIndex.h" //For layer index settings.
#include "types/Ratio.h" //For ratio settings and percentages.
#include "types/Temperature.h" //For temperature settings.
#include "types/Velocity.h" //For velocity settings.
#include "../Application.h" //To get the extruders.
#include "../ExtruderTrain.h"
#include "../Slice.h"
#include "../utils/floatpoint.h" //For FMatrix3x3.
#include "../utils/logoutput.h"
#include "../utils/string.h" //For Escaped.

namespace cura
{

Settings::Settings()
{
    parent = nullptr; //Needs to be properly initialised because we check against this if the parent is not set.
}

void Settings::add(const std::string& key, const std::string value)
{
    if (settings.find(key) != settings.end()) //Already exists.
    {
        settings[key] = value;
    }
    else //New setting.
    {
        settings.emplace(key, value);
    }
}

template<> std::string Settings::get<std::string>(const std::string& key) const
{
    //If this settings base has a setting value for it, look that up.
    if (settings.find(key) != settings.end())
    {
        return settings.at(key);
    }

    const std::unordered_map<std::string, ExtruderTrain*>& limit_to_extruder = Application::getInstance().current_slice->scene.limit_to_extruder;
    if (limit_to_extruder.find(key) != limit_to_extruder.end())
    {
        return limit_to_extruder.at(key)->settings.getWithoutLimiting(key);
    }

    if (parent)
    {
        return parent->get<std::string>(key);
    }

    logError("Trying to retrieve setting with no value given: '%s'\n", key.c_str());
    std::exit(2);
}

template<> double Settings::get<double>(const std::string& key) const
{
    return atof(get<std::string>(key).c_str());
}

template<> size_t Settings::get<size_t>(const std::string& key) const
{
    return std::stoul(get<std::string>(key).c_str());
}

template<> bool Settings::get<bool>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "on" || value == "yes" || value == "true" || value == "True")
    {
        return true;
    }
    const int num = atoi(value.c_str());
    return num != 0;
}

template<> ExtruderTrain& Settings::get<ExtruderTrain&>(const std::string& key) const
{
    int extruder_nr = std::atoi(get<std::string>(key).c_str());
    if (extruder_nr < 0)
    {
        extruder_nr = get<size_t>("extruder_nr");
    }
    return Application::getInstance().current_slice->scene.extruders[extruder_nr];
}

template<> LayerIndex Settings::get<LayerIndex>(const std::string& key) const
{
    return std::atoi(get<std::string>(key).c_str()) - 1; //For the user we display layer numbers starting from 1, but we start counting from 0. Still it may be negative for Raft layers.
}

template<> coord_t Settings::get<coord_t>(const std::string& key) const
{
    return MM2INT(get<double>(key)); //The settings are all in millimetres, but we need to interpret them as microns.
}

template<> AngleRadians Settings::get<AngleRadians>(const std::string& key) const
{
    return get<double>(key) * M_PI / 180; //The settings are all in degrees, but we need to interpret them as radians.
}

template<> AngleDegrees Settings::get<AngleDegrees>(const std::string& key) const
{
    return get<double>(key);
}

template<> Temperature Settings::get<Temperature>(const std::string& key) const
{
    return get<double>(key);
}

template<> Velocity Settings::get<Velocity>(const std::string& key) const
{
    return get<double>(key);
}

template<> Ratio Settings::get<Ratio>(const std::string& key) const
{
    return get<double>(key) / 100.0; //The settings are all in percentages, but we need to interpret them as radians.
}

template<> Duration Settings::get<Duration>(const std::string& key) const
{
    return get<double>(key);
}

template<> DraftShieldHeightLimitation Settings::get<DraftShieldHeightLimitation>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "limited")
    {
        return DraftShieldHeightLimitation::LIMITED;
    }
    else //if (value == "full") or default.
    {
        return DraftShieldHeightLimitation::FULL;
    }
}

template<> FlowTempGraph Settings::get<FlowTempGraph>(const std::string& key) const
{
    std::string value_string = get<std::string>(key);

    FlowTempGraph result;
    if (value_string.empty())
    {
        return result; //Empty at this point.
    }
    /* Match with:
     * - the last opening bracket '['
     * - then a bunch of characters until the first comma
     * - a comma
     * - a bunch of characters until the first closing bracket ']'.
     * This matches with any substring which looks like "[ 124.512 , 124.1 ]".
     */
    std::regex regex("(\\[([^,\\[]*),([^,\\]]*)\\])");

    //Default constructor = end-of-sequence:
    std::regex_token_iterator<std::string::iterator> rend;

    const int submatches[] = {1, 2, 3}; //Match whole pair, first number and second number of a pair.
    std::regex_token_iterator<std::string::iterator> match_iter(value_string.begin(), value_string.end(), regex, submatches);
    while (match_iter != rend)
    {
        match_iter++; //Match the whole pair.
        if (match_iter == rend)
        {
            break;
        }
        std::string first_substring = *match_iter++;
        std::string second_substring = *match_iter++;
        try
        {
            double first = std::stod(first_substring);
            double second = std::stod(second_substring);
            result.data.emplace_back(first, second);
        }
        catch (const std::invalid_argument& e)
        {
            logError("Couldn't read 2D graph element [%s,%s] in setting '%s'. Ignored.\n", first_substring.c_str(), second_substring.c_str(), key.c_str());
        }
    }

    return result;
}

template<> FMatrix3x3 Settings::get<FMatrix3x3>(const std::string& key) const
{
    const std::string value_string = get<std::string>(key);

    FMatrix3x3 result;
    if (value_string.empty())
    {
        return result; //Standard matrix ([[1,0,0], [0,1,0], [0,0,1]]).
    }

    std::string num("([^,\\] ]*)"); //Match with anything but the next ',' ']' or space and capture the match.
    std::ostringstream row; //Match with "[num,num,num]" and ignore whitespace.
    row << "\\s*\\[\\s*" << num << "\\s*,\\s*" << num << "\\s*,\\s*" << num << "\\s*\\]\\s*";
    std::ostringstream matrix; //Match with "[row,row,row]" and ignore whitespace.
    matrix << "\\s*\\[\\s*" << row.str() << "\\s*,\\s*" << row.str() << "\\s*,\\s*" << row.str() << "\\]\\s*";

    std::regex point_matrix_regex(matrix.str());
    std::cmatch sub_matches; //Same as std::match_results<const char*> cm;
    std::regex_match(value_string.c_str(), sub_matches, point_matrix_regex);
    if (sub_matches.size() != 10) //One match for the whole string, nine for the cells.
    {
        logWarning("Mesh transformation matrix could not be parsed!\n\tFormat should be [[f,f,f], [f,f,f], [f,f,f]] allowing whitespace anywhere in between.\n\tWhile what was given was \"%s\".\n", value_string.c_str());
        return result; //Standard matrix ([[1,0,0], [0,1,0], [0,0,1]]).
    }

    size_t sub_match_index = 1; //Skip the first because the first submatch is the whole string.
    for (size_t x = 0; x < 3; x++)
    {
        for (size_t y = 0; y < 3; y++)
        {
            std::sub_match<const char*> sub_match = sub_matches[sub_match_index];
            result.m[y][x] = strtod(std::string(sub_match.str()).c_str(), nullptr);
            sub_match_index++;
        }
    }

    return result;
}

template<> FQuaternion Settings::get<FQuaternion>(const std::string& key) const
{
    const std::string value_string = get<std::string>(key);

    FQuaternion result;
    if (value_string.empty())
    {
        return result; //Standard matrix ([[1,0,0,0], [0,1,0,0], [0,0,1,0], [0,0,0,1]]).
    }

    std::string num("([^,\\] ]*)"); //Match with anything but the next ',' ']' or space and capture the match.
    std::ostringstream row; //Match with "[num,num,num,num]" and ignore whitespace.
    row << "\\s*\\[\\s*" << num << "\\s*,\\s*" << num << "\\s*,\\s*" << num << "\\s*,\\s*" << num << "\\s*\\]\\s*";
    std::ostringstream matrix; //Match with "[row,row,row,row]" and ignore whitespace.
    matrix << "\\s*\\[\\s*" << row.str() << "\\s*,\\s*" << row.str() << "\\s*,\\s*" << row.str() << "\\s*,\\s*" << row.str() << "\\]\\s*";

    std::regex point_matrix_regex(matrix.str());
    std::cmatch sub_matches; //Same as std::match_results<const char*> cm;
    std::regex_match(value_string.c_str(), sub_matches, point_matrix_regex);
    if (sub_matches.size() != 17) //One match for the whole string, 16 for the cells.
    {
        logWarning("Mesh transformation matrix could not be parsed!\n\tFormat should be [[f,f,f,f], [f,f,f,f], [f,f,f,f], [f,,ff,f]] allowing whitespace anywhere in between.\n\tWhile what was given was \"%s\".\n", value_string.c_str());
        return result; //Standard matrix ([[1,0,0,0], [0,1,0,0], [0,0,1,0], [0,0,0,1]]).
    }

    size_t sub_match_index = 1; //Skip the first because the first submatch is the whole string.
    for (size_t row = 0; row < 4; row++)
    {
        for (size_t col = 0; col < 4; col++)
        {
            std::sub_match<const char*> sub_match = sub_matches[sub_match_index];
            result.m[row][col] = strtod(std::string(sub_match.str()).c_str(), nullptr);
            sub_match_index++;
        }
    }

    return result;
}

template<> EGCodeFlavor Settings::get<EGCodeFlavor>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    //I wish that switch statements worked for std::string...
    if (value == "Griffin")
    {
        return EGCodeFlavor::GRIFFIN;
    }
    else if (value == "UltiGCode")
    {
        return EGCodeFlavor::ULTIGCODE;
    }
    else if (value == "Makerbot")
    {
        return EGCodeFlavor::MAKERBOT;
    }
    else if (value == "BFB")
    {
        return EGCodeFlavor::BFB;
    }
    else if (value == "MACH3")
    {
        return EGCodeFlavor::MACH3;
    }
    else if (value == "RepRap (Volumetric)")
    {
        return EGCodeFlavor::MARLIN_VOLUMATRIC;
    }
    else if (value == "Repetier")
    {
        return EGCodeFlavor::REPETIER;
    }
    else if (value == "RepRap (RepRap)")
    {
        return EGCodeFlavor::REPRAP;
    }
    else if (value == "Picaso")
    {
        return EGCodeFlavor::PICASO;
    }
    //Default:
    return EGCodeFlavor::MARLIN;
}

template<> EFillMethod Settings::get<EFillMethod>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "lines")
    {
        return EFillMethod::LINES;
    }
    else if (value == "grid")
    {
        return EFillMethod::GRID;
    }
    else if (value == "cubic")
    {
        return EFillMethod::CUBIC;
    }
    else if (value == "cubicsubdiv")
    {
        return EFillMethod::CUBICSUBDIV;
    }
    else if (value == "tetrahedral")
    {
        return EFillMethod::TETRAHEDRAL;
    }
    else if (value == "quarter_cubic")
    {
        return EFillMethod::QUARTER_CUBIC;
    }
    else if (value == "triangles")
    {
        return EFillMethod::TRIANGLES;
    }
    else if (value == "trihexagon")
    {
        return EFillMethod::TRIHEXAGON;
    }
    else if (value == "concentric")
    {
        return EFillMethod::CONCENTRIC;
    }
    else if (value == "zigzag")
    {
        return EFillMethod::ZIG_ZAG;
    }
    else if (value == "cross")
    {
        return EFillMethod::CROSS;
    }
    else if (value == "cross_3d")
    {
        return EFillMethod::CROSS_3D;
    }
    else if (value == "gyroid")
    {
        return EFillMethod::GYROID;
    }
	else if (value == "honeycomb_full")
	{
		return EFillMethod::HONEYCOMB_FULL;
	}
	else if (value == "honeycomb_simple")
	{
		return EFillMethod::HONEYCOMB_SIMPLE;
	}
    else //Default.
    {
        return EFillMethod::NONE;
    }
}

template<> EPlatformAdhesion Settings::get<EPlatformAdhesion>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "brim")
    {
        return EPlatformAdhesion::BRIM;
    }
    else if (value == "raft")
    {
        return EPlatformAdhesion::RAFT;
    }
    else if (value == "none")
    {
        return EPlatformAdhesion::NONE;
    }
    else //Default.
    {
        return EPlatformAdhesion::SKIRT;
    }
}

template<> ESupportType Settings::get<ESupportType>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "everywhere")
    {
        return ESupportType::EVERYWHERE;
    }
    else if (value == "buildplate")
    {
        return ESupportType::PLATFORM_ONLY;
    }
    else //Default.
    {
        return ESupportType::NONE;
    }
}

template<> EZSeamType Settings::get<EZSeamType>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "random")
    {
        return EZSeamType::RANDOM;
    }
    else if (value == "back") //It's called 'back' internally because originally this was intended to allow the user to put the seam in the back of the object where it's less visible.
    {
        return EZSeamType::USER_SPECIFIED;
    }
    else if (value == "sharpest_corner")
    {
        return EZSeamType::SHARPEST_CORNER;
    }
    else //Default.
    {
        return EZSeamType::SHORTEST;
    }
}

template<> EZSeamCornerPrefType Settings::get<EZSeamCornerPrefType>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "z_seam_corner_inner")
    {
        return EZSeamCornerPrefType::Z_SEAM_CORNER_PREF_INNER;
    }
    else if (value == "z_seam_corner_outer")
    {
        return EZSeamCornerPrefType::Z_SEAM_CORNER_PREF_OUTER;
    }
    else if (value == "z_seam_corner_any")
    {
        return EZSeamCornerPrefType::Z_SEAM_CORNER_PREF_ANY;
    }
    else if (value == "z_seam_corner_weighted")
    {
        return EZSeamCornerPrefType::Z_SEAM_CORNER_PREF_WEIGHTED;
    }
	else if (value == "z_seam_corner_smooth")
	{
		return EZSeamCornerPrefType::Z_SEAM_CORNER_PREF_SMOOTH;
	}
    else //Default.
    {
        return EZSeamCornerPrefType::Z_SEAM_CORNER_PREF_NONE;
    }
}

template<> EZSeamCross Settings::get<EZSeamCross>(const std::string& key) const
{
	const std::string& value = get<std::string>(key);
	if (value == "z_seam_cross_normal")
	{
		return EZSeamCross::NORMAL;
	}
	else if (value == "z_seam_cross_half")
	{
		return EZSeamCross::HALF;
	}
	else if (value == "z_seam_cross_quarter")
	{
		return EZSeamCross::QUARTER;
	}
	else //Default "z_seam_cross_none"
	{
		return EZSeamCross::NONE;
	}
}

template<> ESurfaceMode Settings::get<ESurfaceMode>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "surface")
    {
        return ESurfaceMode::SURFACE;
    }
    else if (value == "both")
    {
        return ESurfaceMode::BOTH;
    }
    else //Default.
    {
        return ESurfaceMode::NORMAL;
    }
}

template<> FillPerimeterGapMode Settings::get<FillPerimeterGapMode>(const std::string& key) const
{
    if (get<std::string>(key) == "everywhere")
    {
        return FillPerimeterGapMode::EVERYWHERE;
    }
    else //Default.
    {
        return FillPerimeterGapMode::NOWHERE;
    }
}

template<> BuildPlateShape Settings::get<BuildPlateShape>(const std::string& key) const
{
    if (get<std::string>(key) == "elliptic")
    {
        return BuildPlateShape::ELLIPTIC;
    }
    else //Default.
    {
        return BuildPlateShape::RECTANGULAR;
    }
}

template<> CombingMode Settings::get<CombingMode>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "off")
    {
        return CombingMode::OFF;
    }
    else if (value == "noskin")
    {
        return CombingMode::NO_SKIN;
    }
    else if (value == "infill")
    {
        return CombingMode::INFILL;
    }
    else //Default.
    {
        return CombingMode::ALL;
    }
}

template<> SupportDistPriority Settings::get<SupportDistPriority>(const std::string& key) const
{
    if (get<std::string>(key) == "z_overrides_xy")
    {
        return SupportDistPriority::Z_OVERRIDES_XY;
    }
    else //Default.
    {
        return SupportDistPriority::XY_OVERRIDES_Z;
    }
}

template<> SlicingTolerance Settings::get<SlicingTolerance>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "inclusive")
    {
        return SlicingTolerance::INCLUSIVE;
    }
    else if (value == "exclusive")
    {
        return SlicingTolerance::EXCLUSIVE;
    }
    else //Default.
    {
        return SlicingTolerance::MIDDLE;
    }
}

template<> MagicPrintOrder Settings::get<MagicPrintOrder>(const std::string& key) const
{
    const std::string& value = get<std::string>(key);
    if (value == "support")
    {
        return MagicPrintOrder::SUPPORT;
    }
    if (value == "model")
    {
        return MagicPrintOrder::MODEL;
    }
    return MagicPrintOrder::SUPPORT;
}

template<> std::vector<double> Settings::get<std::vector<double>>(const std::string& key) const
{
    const std::string& value_string = get<std::string>(key);

    std::vector<double> result;
    if (value_string.empty())
    {
        return result;
    }

    /* We're looking to match one or more floating point values separated by
     * commas and surrounded by square brackets. Note that because the QML
     * RegExpValidator only stops unrecognised characters being input and
     * doesn't actually barf if the trailing ']' is missing, we are lenient here
     * and make that bracket optional.
     */
    std::regex list_contents_regex(R"(\[([^\]]*)\]?)");
    std::smatch list_contents_match;
    if (std::regex_search(value_string, list_contents_match, list_contents_regex) && list_contents_match.size() > 1)
    {
        std::string elements = list_contents_match.str(1);
        std::regex element_regex(R"(\s*([+-]?[0-9]*\.?[0-9]+)\s*,?)");
        std::regex_token_iterator<std::string::iterator> rend; //Default constructor gets the end-of-sequence iterator.

        std::regex_token_iterator<std::string::iterator> match_iter(elements.begin(), elements.end(), element_regex, 0);
        while (match_iter != rend)
        {
            std::string value = *match_iter++;
            try
            {
                result.push_back(std::stod(value));
            }
            catch (const std::invalid_argument& e)
            {
                logError("Couldn't read floating point value (%s) in setting '%s'. Ignored.\n", value.c_str(), key.c_str());
            }
        }
    }
    return result;
}

template<> std::vector<int> Settings::get<std::vector<int>>(const std::string& key) const
{
    std::vector<double> values_doubles = get<std::vector<double>>(key);
    std::vector<int> values_ints;
    values_ints.reserve(values_doubles.size());
    for (double value : values_doubles)
    {
        values_ints.push_back(std::round(value)); //Round to nearest integer.
    }
    return values_ints;
}

template<> std::vector<AngleDegrees> Settings::get<std::vector<AngleDegrees>>(const std::string& key) const
{
    std::vector<double> values_doubles = get<std::vector<double>>(key);
    return std::vector<AngleDegrees>(values_doubles.begin(), values_doubles.end()); //Cast them to AngleDegrees.
}

const std::string Settings::getAllSettingsString() const
{
    std::stringstream sstream;
    for (const std::pair<std::string, std::string> pair : settings)
    {
        char buffer[4096];
        snprintf(buffer, 4096, " -s %s=\"%s\"", pair.first.c_str(), Escaped{pair.second.c_str()}.str);
        sstream << buffer;
    }
    return sstream.str();
}

bool Settings::has(const std::string& key) const
{
    return settings.find(key) != settings.end();
}

void Settings::setParent(Settings* new_parent)
{
    parent = new_parent;
}

std::string Settings::getWithoutLimiting(const std::string& key) const
{
    if (settings.find(key) != settings.end())
    {
        return settings.at(key);
    }
    else if (parent)
    {
        return parent->get<std::string>(key);
    }
    else
    {
        logError("Trying to retrieve setting with no value given: '%s'\n", key.c_str());
        std::exit(2);
    }
}

}//namespace cura

